GH-1522 TerraWorkspaceSink: identifier can be a workflow input

okotsopoulos opened the pull request:
### Purpose
<!-- Please explain the purpose of this PR and include links to any ticket that it fixes: -->

- https://broadinstitute.atlassian.net/browse/GH-1522

The TerraWorkspaceSink requires that an identifier be specified.  Previously, it had to be "the name of a pipeline output that should be used as the name of each newly created entity."

For eMerge arrays and imputation workflows, we've found that the most sensible identifiers are actually pipeline inputs.  In such cases, Lantern had to redundantly thread pipeline inputs through as outputs.  This is a common-enough use case that it was worth fixing WFL-side.

### Changes
<!-- Please list out what major changes were made in this PR to address the issue: -->

Where can identifiers come from?
- Old behavior: workflow outputs only.
- New behavior: workflow outputs, but if no match fall back to workflow inputs.

What happens if there's no match for an identifier in a workflow and `entityName = nil`?
- Old behavior: try to sink the workflow outputs but fail with `400 Bad Request` from Rawls/Firecloud.
- New behavior: throw a descriptive exception, do not try to sink the workflow outputs.

I also added new unit tests and supplemented an existing integration test.

Documentation changes handled separately: https://github.com/broadinstitute/wfl/pull/535

### Manual Testing

For an eMerge Imputation workload with a bad `sink.identifier`, here's what's thrown when `sink/update-terra-workspace-sink` runs:

```
{"severity":"ERROR","logging.googleapis.com/sourceLocation":{"file":"wfl/server.clj","line":103},"timestamp":"2021-11-05T16:26:49.619452Z","message":"clojure.lang.ExceptionInfo: Entity name not found: sink.identifer not present in workflow outputs or inputs {:sink {:id 19, :details \"TerraWorkspaceSink_000000231\", :workspace \"emerge_prod/Imputation_test\", :entityType \"Imputation_Outputs_Table\", :fromOutputs {:imputed_multisample_vcf \"imputed_multisample_vcf\", :failed_chunks \"failed_chunks\", :imputed_single_sample_vcfs \"imputed_single_sample_vcfs\", :aggregated_imputation_metrics \"aggregated_imputation_metrics\", :imputed_single_sample_vcf_indices \"imputed_single_sample_vcf_indices\", :n_failed_chunks \"n_failed_chunks\", :chunks_info \"chunks_info\", :imputed_multisample_vcf_index \"imputed_multisample_vcf_index\"}, :identifier \"output_callset_name_bad\", :type \"TerraWorkspaceSink\"}, :workflow {:inputs {:haplotype_database \"gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.haplotype_database.txt\", :AggregateImputationQCMetrics.cpu 1, :FindSitesUniqueToFileTwoOnly.memory_mb 4000, :SetIdsVcfToImpute.cpu 1, :SetIdsVcfToImpute.memory_mb 4000, :Minimac4.memory_mb 4000, :SetIdsVcfToImpute.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :merge_ssvcf_mem_mb 3000, :reference_panel_path \"gs://gcp-public-data--broad-references/hg19/v0/1000G_reference_panel/\", :RemoveAnnotations.disk_size_gb nil, :ref_dict \"gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.dict\", :MergeImputationQCMetrics.rtidyverse_docker \"rocker/tidyverse:4.1.0\", :MergeSingleSampleVcfs.cpu 1, :SplitMultiSampleVcf.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :CountSamples.memory_mb 3000, :OptionalQCSites.cpu 1, :MergeImputationQCMetrics.memory_mb 2000, :MergeSingleSampleVcfs.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :FindSitesUniqueToFileTwoOnly.disk_size_gb nil, :CheckChunks.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :CountSamples.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :optional_qc_max_missing 0.05, :OptionalQCSites.memory_mb 16000, :ExtractIdsVcfToImpute.cpu 1, :StoreChunksInfo.disk_size_gb 10, :multi_sample_vcf_index nil, :CountSamples.disk_size_gb nil, :SeparateMultiallelics.disk_size_gb nil, :single_sample_vcf_indices [\"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_61696bd5-7a01-4d3e-8f84-e6f31bb29581\" \"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_7ffbbbf7-d071-4f55-9601-f10c63e33232\" \"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_2778dfb3-4ac5-4a09-9409-7af0694d7909\" \"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_322f2a28-7b11-45dc-8bfc-20ef88d35d44\"], :MergeImputationQCMetrics.disk_size_gb nil, :ExtractIDs.cpu 1, :ExtractIdsVcfToImpute.memory_mb 4000, :RemoveSymbolicAlleles.cpu 1, :SeparateMultiallelics.cpu 1, :PhaseVariantsEagle.cpu 8, :MergeImputationQCMetrics.cpu 1, :CalculateChromosomeLength.ubuntu_docker \"ubuntu:20.04\", :CheckChunks.cpu 1, :Minimac4.cpu 1, :InterleaveVariants.disk_size_gb nil, :CalculateChromosomeLength.memory_mb 2000, :Minimac4.disk_size_gb nil, :SplitMultiSampleVcf.memory_mb 8000, :multi_sample_vcf nil, :UpdateHeader.memory_mb 8000, :InterleaveVariants.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :GenerateChunk.cpu 1, :ExtractIDs.memory_mb 4000, :frac_well_imputed_threshold 0.9, :InterleaveVariants.memory_mb 16000, :GatherVcfs.disk_size_gb nil, :RemoveAnnotations.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :CalculateChromosomeLength.disk_size_gb nil, :SetIdsVcfToImpute.disk_size_gb nil, :RemoveAnnotations.memory_mb 3000, :GenerateChunk.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :single_sample_vcfs [\"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_ed719b05-9e28-491b-84ef-975aee7dcdaf\" \"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_77d7659a-79ba-47e7-a1bd-993793c7786f\" \"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_bf2df1d4-e33f-44cf-b1b6-fa90ffe076bd\" \"drs://data.terra.bio/v1_97ef7104-d9d0-4a99-9cbf-0eb0591bcec1_38578750-b73a-4618-aa6b-38e6e255620d\"], :contigs [\"21\" \"22\"], :bcf_suffix \".bcf\", :CalculateChromosomeLength.cpu 1, :CheckChunks.memory_mb 4000, :FindSitesUniqueToFileTwoOnly.cpu 1, :split_output_to_single_sample false, :RemoveAnnotations.cpu 1, :UpdateHeader.disk_size_gb nil, :CountVariantsInChunks.disk_size_gb 100, :FindSitesUniqueToFileTwoOnly.ubuntu_docker \"ubuntu:20.04\", :ExtractIDs.disk_size_gb nil, :SelectVariantsByIds.cpu 1, :PhaseVariantsEagle.eagle_docker \"us.gcr.io/broad-dsde-methods/imputation_eagle_docker:v1.0.0\", :CountVariantsInChunks.cpu 1, :MergeSingleSampleVcfs.disk_size_gb nil, :SetIDs.disk_size_gb nil, :CountVariantsInChunks.memory_mb 4000, :UpdateHeader.cpu 1, :perform_extra_qc_steps false, :InterleaveVariants.cpu 1, :CheckChunks.disk_size_gb nil, :StoreChunksInfo.cpu 1, :vcf_index_suffix \".vcf.gz.tbi\", :output_callset_name \"emerge_test_WFL_identifier_as_input_variable\", :chunkLength 25000000, :ExtractIdsVcfToImpute.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :GenerateChunk.memory_mb 8000, :SetIDs.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :SelectVariantsByIds.disk_size_gb nil, :AggregateImputationQCMetrics.rtidyverse_docker \"rocker/tidyverse:4.1.0\", :UpdateHeader.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :RemoveSymbolicAlleles.disk_size_gb nil, :OptionalQCSites.bcftools_vcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :bcf_index_suffix \".bcf.csi\", :RemoveSymbolicAlleles.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :chunks_fail_threshold 1, :RemoveSymbolicAlleles.memory_mb 4000, :OptionalQCSites.disk_size_gb nil, :SetIDs.memory_mb 4000, :AggregateImputationQCMetrics.disk_size_gb nil, :vcf_suffix \".vcf.gz\", :GatherVcfs.memory_mb 16000, :SeparateMultiallelics.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :SetIDs.cpu 1, :CountVariantsInChunks.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :genetic_maps_eagle \"gs://broad-gotc-test-storage/imputation/eagle_genetic_map/genetic_map_hg19_withX.txt.gz\", :chunkOverlaps 5000000, :ExtractIdsVcfToImpute.disk_size_gb nil, :m3vcf_suffix \".cleaned.m3vcf.gz\", :StoreChunksInfo.memory_mb 2000, :GatherVcfs.cpu 1, :SelectVariantsByIds.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :SeparateMultiallelics.memory_mb 4000, :ExtractIDs.bcftools_docker \"us.gcr.io/broad-dsde-methods/imputation_bcftools_vcftools_docker:v1.0.0\", :SplitMultiSampleVcf.cpu 1, :SplitMultiSampleVcf.disk_size_gb nil, :PhaseVariantsEagle.disk_size_gb nil, :AggregateImputationQCMetrics.memory_mb 2000, :StoreChunksInfo.rtidyverse_docker \"rocker/tidyverse:4.1.0\", :SelectVariantsByIds.memory_mb 16000, :GatherVcfs.gatk_docker \"us.gcr.io/broad-gatk/gatk:4.1.9.0\", :Minimac4.minimac4_docker \"us.gcr.io/broad-dsde-methods/imputation-minimac-docker:v1.0.0\", :CountSamples.cpu 1, :GenerateChunk.disk_size_gb nil, :optional_qc_hwe 1.0E-6, :PhaseVariantsEagle.memory_mb 32000}, :uuid \"1a6a697b-f877-47a4-80fc-674aa29eee0d\", :status \"Succeeded\", :outputs {:imputed_single_sample_vcfs nil, :failed_chunks \"gs://fc-5ac2145c-7c19-4be4-bfb4-b41850e7f4e1/14bffc69-6ce7-4615-b318-7ef1c457c894/Imputation/1a6a697b-f877-47a4-80fc-674aa29eee0d/call-StoreChunksInfo/emerge_test_WFL_identifier_as_input_variable_failed_chunks.tsv\", :imputed_single_sample_vcf_indices nil, :imputed_multisample_vcf \"gs://fc-5ac2145c-7c19-4be4-bfb4-b41850e7f4e1/14bffc69-6ce7-4615-b318-7ef1c457c894/Imputation/1a6a697b-f877-47a4-80fc-674aa29eee0d/call-InterleaveVariants/emerge_test_WFL_identifier_as_input_variable.vcf.gz\", :aggregated_imputation_metrics \"gs://fc-5ac2145c-7c19-4be4-bfb4-b41850e7f4e1/14bffc69-6ce7-4615-b318-7ef1c457c894/Imputation/1a6a697b-f877-47a4-80fc-674aa29eee0d/call-MergeImputationQCMetrics/emerge_test_WFL_identifier_as_input_variable_aggregated_imputation_metrics.tsv\", :n_failed_chunks \"gs://fc-5ac2145c-7c19-4be4-bfb4-b41850e7f4e1/14bffc69-6ce7-4615-b318-7ef1c457c894/Imputation/1a6a697b-f877-47a4-80fc-674aa29eee0d/call-StoreChunksInfo/n_failed_chunks.txt\", :chunks_info \"gs://fc-5ac2145c-7c19-4be4-bfb4-b41850e7f4e1/14bffc69-6ce7-4615-b318-7ef1c457c894/Imputation/1a6a697b-f877-47a4-80fc-674aa29eee0d/call-StoreChunksInfo/emerge_test_WFL_identifier_as_input_variable_chunk_info.tsv\", :imputed_multisample_vcf_index \"gs://fc-5ac2145c-7c19-4be4-bfb4-b41850e7f4e1/14bffc69-6ce7-4615-b318-7ef1c457c894/Imputation/1a6a697b-f877-47a4-80fc-674aa29eee0d/call-InterleaveVariants/emerge_test_WFL_identifier_as_input_variable.vcf.gz.tbi\"}, :updated #inst \"2021-11-05T16:06:15.503727000-00:00\"}}"}
```

When setting `sink.identifier` to a workflow input not present in the workflow's outputs:
```
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":189},"timestamp":"2021-11-05T16:33:15.912004Z","message":"coercing workflow 1a6a697b-f877-47a4-80fc-674aa29eee0d outputs to Imputation_Outputs_Table"}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":198},"timestamp":"2021-11-05T16:33:16.700403Z","message":"entity emerge_test_WFL_identifier_as_input_variable exists - deleting previous entity."}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":200},"timestamp":"2021-11-05T16:33:17.188490Z","message":"upserting workflow 1a6a697b-f877-47a4-80fc-674aa29eee0d outputs as emerge_test_WFL_identifier_as_input_variable"}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/executor.clj","line":329},"timestamp":"2021-11-05T16:33:17.856191Z","message":"jdbc/query: okotsopo@jdbc:postgresql:wfl#734 SELECT * FROM TerraExecutor_000000231\n               WHERE consumed IS NULL\n               AND   status = 'Succeeded'\n               ORDER BY id ASC\n               LIMIT 1"}
{"severity":"INFO","logging.googleapis.com/sourceLocation":{"file":"wfl/executor.clj","line":371},"timestamp":"2021-11-05T16:33:17.863614Z","message":"jdbc/update! okotsopo@jdbc:postgresql:wfl#734 TerraExecutor_000000231 {:consumed #object[java.time.OffsetDateTime 0x20cde482 \"2021-11-05T16:33:17.863349Z\"], :updated #object[java.time.OffsetDateTime 0x20cde482 \"2021-11-05T16:33:17.863349Z\"]} [\"id = ?\" 2]"}
{"severity":"INFO","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":203},"timestamp":"2021-11-05T16:33:17.866055Z","message":"sunk workflow 1a6a697b-f877-47a4-80fc-674aa29eee0d to emerge_prod/Imputation_test as emerge_test_WFL_identifier_as_input_variable"}
{"severity":"INFO","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":205},"timestamp":"2021-11-05T16:33:17.871406Z","message":"jdbc/insert okotsopo@jdbc:postgresql:wfl#734 TerraWorkspaceSink_000000231 {:entity \"emerge_test_WFL_identifier_as_input_variable\", :updated #object[java.time.OffsetDateTime 0x6c0a8eaa \"2021-11-05T16:33:17.871183Z\"], :workflow \"1a6a697b-f877-47a4-80fc-674aa29eee0d\"}"}
{"severity":"INFO","logging.googleapis.com/sourceLocation":{"file":"wfl/module/covid.clj","line":49},"timestamp":"2021-11-05T16:33:17.880820Z","message":"jdbc/update! okotsopo@jdbc:postgresql:wfl#734 :workload {:updated #object[java.time.OffsetDateTime 0x3c279037 \"2021-11-05T16:33:10.125489Z\"]} [\"id = ?\" 231]"}
```

The workflow in question:
https://job-manager.dsde-prod.broadinstitute.org/jobs/1a6a697b-f877-47a4-80fc-674aa29eee0d

And its written outputs:
<img width="1788" alt="Screen Shot 2021-11-05 at 3 13 45 PM" src="https://user-images.githubusercontent.com/79769153/140565982-282ed5ce-9205-403a-a459-c34f88c8a1df.png">

### System Tests

I ran them locally and they passed, save for a known unrelated failure:

```
wm111-e35:wfl okotsopo$ make api TARGET=system
export CPCACHE=/Users/okotsopo/wfl/api/.cpcache;            \
	export WFL_WFL_URL=http://localhost:3000; \
	clojure  -M:parallel-test wfl.system.v1-endpoint-test | \
	tee /Users/okotsopo/wfl/derived/api/system.log
WARNING: Specified path is external to project: ../derived/api/src
WARNING: Specified path is external to project: ../derived/api/resources

ERROR in (test-workload-sink-outputs-to-tdr) (support.clj:201)
Uncaught exception, not in assertion.
expected: nil
  actual: clojure.lang.ExceptionInfo: clj-http: status 401
{:cached nil, :request-time 191, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :http-client #object[org.apache.http.impl.client.InternalHttpClient 0xa00ec06 "org.apache.http.impl.client.InternalHttpClient@a00ec06"], :chunked? true, :type :clj-http.client/unexceptional-status, :reason-phrase "", :headers {"Access-Control-Allow-Headers" "DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin", "Server" "Apache", "Via" "1.1 google", "Content-Type" "application/json", "X-Content-Type-Options" "nosniff", "Alt-Svc" "clear", "X-Frame-Options" "SAMEORIGIN", "Strict-Transport-Security" "max-age=31536000;includeSubDomains", "Connection" "close", "Transfer-Encoding" "chunked", "Access-Control-Allow-Methods" "GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD", "Date" "Fri, 05 Nov 2021 21:24:00 GMT", "Vary" "Origin"}, :orig-content-encoding nil, :status 401, :length -1, :body "{\"message\":\"Accessing bucket broad-gotc-dev-wfl-ptc-test-inputs is not authorized. Please be sure to grant \\\"Storage Object Viewer\\\" permissions to the TDR service account and your Terra proxy user group\",\"errorDetail\":[]}", :trace-redirects []}
```

tbl3rd commented on api/test/wfl/integration/sinks/workspace_sink_test.clj:
```diff
@@ -79,44 +80,59 @@
+           #'sink/entity-exists?       (partial throw-if-called "sink/entity-exists?")
+           #'firecloud/delete-entities (partial throw-if-called "firecloud/delete-entities")}
+          #(is (thrown-with-msg?
+                ExceptionInfo (re-pattern sink/entity-name-not-found-error-message)
```
If you took my prior suggestion, then this should be:
```clojure
ExceptionInfo (-> sink/entity-name-not-found-error-message
                  var var-get re-pattern)
```
or something.

okotsopoulos replied:
(Doesn't really matter but) wouldn't this work?

`ExceptionInfo (re-pattern (sink/entity-name-not-found-error-message))`

okotsopoulos replied:
Whoops, bad reading comprehension.  Disregard.

tbl3rd commented on api/test/wfl/integration/sinks/workspace_sink_test.clj:
```diff
@@ -79,44 +80,59 @@
+            (is record "The record was not written to the database")
+            (is (empty? rest) "More than one record was written")
+            (is (= (:uuid workflow) (:workflow record)) "The workflow UUID was not written")
+            (is (= testing-entity-name (:entity record)) "The entity was not correct")))))))
```
Maybe break out he `letfn`s and simplify with `with-redefs`?
```clojure
(defn make-sink
  "Create and load a sink with identifier."
  [identifier]
  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
    (->> {:name           "Terra Workspace"
          :workspace      "workspace-ns/workspace-name"
          :entityType     testing-entity-type
          :fromOutputs    (resources/read-resource
                           "sarscov2_illumina_full/entity-from-outputs.edn")
          :identifier     identifier
          :skipValidation true}
         (sink/create-sink! tx (rand-int 1000000))
         (zipmap [:sink_type :sink_items])
         (sink/load-sink! tx))))

(defn verify-upsert-request
  "Mock rawls/batch-upsert to validate ENTITIES."
  [workspace [[type name _] :as _entities]]
  (is (= "workspace-ns/workspace-name" workspace))
  (is (= testing-entity-type type))
  (is (= testing-entity-name name )))

(deftest test-update-terra-workspace-sink
  (let [workflow {:uuid    "2768b29e-c808-4bd6-a46b-6c94fd2a67aa"
                  :status  "Succeeded"
                  :outputs (-> "sarscov2_illumina_full/outputs.edn"
                               resources/read-resource
                               (assoc :flowcell_id testing-entity-name))}
        executor (make-queue-from-list [[nil workflow]])
        throws   (make-sink "not-a-workflow-input-or-output")
        updates  (make-sink "flowcell_id")]
    (testing "Sink identifier matches no workflow output or input"
      (with-redefs
        [rawls/batch-upsert        #(throw (Throwable. "rawls/batch-upsert"))
         sink/entity-exists?       #(throw (Throwable. "sink/entity-exists?"))
         firecloud/delete-entities #(throw (Throwable. "firecloud/delete-entities"))]
        (is (thrown-with-msg?
             ExceptionInfo (-> sink/entity-name-not-found-error-message
                               var var-get re-pattern)
             (sink/update-sink! executor throws))
            "Sink update should throw if identifier not found in workflow output or input"))
      (is (== 1 (stage/queue-length executor)) "The workflow should not have been consumed")
      (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
        (let [records (->> throws :details (postgres/get-table tx))]
          (is (empty? records) "No sink records should have been written"))))
    (testing "Sink identifier matches workflow output"
      (with-redefs
        [rawls/batch-upsert        verify-upsert-request
         sink/entity-exists?       (constantly false)
         firecloud/delete-entities #(throw (Throwable. "firecloud/delete-entities"))]
        (sink/update-sink! executor updates))
      (is (zero? (stage/queue-length executor)) "The workflow was not consumed")
      (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
        (let [[record & rest] (->> updates :details (postgres/get-table tx))]
          (is record "The record was not written to the database")
          (is (empty? rest) "More than one record was written")
          (is (= (:uuid workflow) (:workflow record)) "The workflow UUID was not written")
          (is (= testing-entity-name (:entity record)) "The entity was not correct"))))))
```

okotsopoulos replied:
Thanks.

tbl3rd commented on api/test/wfl/unit/sink_test.clj:
```diff
@@ -39,3 +39,22 @@
+    (is (= "cOutput" (#'sink/entity-name-from-workflow workflow "c"))
+        "Key c only present in outputs map")
+    (is (nil? (#'sink/entity-name-from-workflow workflow "d"))
+        "Key d not found in inputs or outputs maps should return nil")))
```
Thanks. Wish I'd read this first.

okotsopoulos replied:
Next time I can call out any test usage examples in the review instructions.

tbl3rd reviewed (approved):
Sorry. This is kind of a surface read because I barely understand it.

okotsopoulos commented on api/src/wfl/sink.clj:
```diff
@@ -171,21 +171,37 @@
-        (jdbc/insert! tx details {:entity   name
+        (jdbc/insert! tx details {:entity   entityName
                                   :updated  (util/utc-now)
                                   :workflow uuid})))))
```
Thanks.

I thought that fussing with existing logging too much would be out of scope, but my preference is logging within the methods called rather than before/after calling them.  Will take a crack at it.

okotsopoulos commented:
System tests succeed outside of a known unrelated failure:

```
$ make api TARGET=system
export CPCACHE=/Users/okotsopo/wfl/api/.cpcache;            \
	export WFL_WFL_URL=http://localhost:3000; \
	clojure  -M:parallel-test wfl.system.v1-endpoint-test | \
	tee /Users/okotsopo/wfl/derived/api/system.log
WARNING: Specified path is external to project: ../derived/api/src
WARNING: Specified path is external to project: ../derived/api/resources

ERROR in (test-workload-sink-outputs-to-tdr) (support.clj:201)
Uncaught exception, not in assertion.
expected: nil
  actual: clojure.lang.ExceptionInfo: clj-http: status 401
{:cached nil, :request-time 284, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :http-client #object[org.apache.http.impl.client.InternalHttpClient 0x13b29e3e "org.apache.http.impl.client.InternalHttpClient@13b29e3e"], :chunked? true, :type :clj-http.client/unexceptional-status, :reason-phrase "", :headers {"Access-Control-Allow-Headers" "DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin", "Server" "Apache", "Via" "1.1 google", "Content-Type" "application/json", "X-Content-Type-Options" "nosniff", "Alt-Svc" "clear", "X-Frame-Options" "SAMEORIGIN", "Strict-Transport-Security" "max-age=31536000;includeSubDomains", "Connection" "close", "Transfer-Encoding" "chunked", "Access-Control-Allow-Methods" "GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD", "Date" "Wed, 10 Nov 2021 00:06:11 GMT", "Vary" "Origin"}, :orig-content-encoding nil, :status 401, :length -1, :body "{\"message\":\"Accessing bucket broad-gotc-dev-wfl-ptc-test-inputs is not authorized for user wfl-non-prod@broad-gotc-dev.iam.gserviceaccount.com. Please be sure to grant \\\"Storage Object Viewer\\\" permissions to the TDR service account and your Terra proxy user group (PROXY_26133376970534d9dddd1@firecloud.org)\",\"errorDetail\":[]}", :trace-redirects []}
```

Before merging, I am waiting on a workload to complete and its outputs sunk to the workspace to verify my logging changes:

https://app.terra.bio/#workspaces/emerge_prod/Imputation_test/job_history/d391c233-c1e7-48a1-b626-e30f1d518b4b

okotsopoulos commented:
Logging changes, sinking workflow outputs:

```
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":196},"timestamp":"2021-11-10T16:57:55.018465Z","message":{"action":"Attempting to sink workflow outputs","workflow":"668e3257-bf67-41cc-9e77-8be40643c521","entityType":"Imputation_Outputs_Table”}}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/service/firecloud.clj","line":158},"timestamp":"2021-11-10T16:57:55.429886Z","message":{"action":"Deleting entities","workspace":"emerge_prod/Imputation_test","entities":[["Imputation_Outputs_Table","emerge_test_WFL_identifier_as_input_variable"]]}}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/service/rawls.clj","line":124},"timestamp":"2021-11-10T16:57:55.912636Z","message":{"action":"Upserting entities","workspace":"emerge_prod/Imputation_test","entities":[["Imputation_Outputs_Table","emerge_test_WFL_identifier_as_input_variable"]]}}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/executor.clj","line":344},"timestamp":"2021-11-10T16:57:56.350520Z","message":"jdbc/query: okotsopo@jdbc:postgresql:wfl#612 SELECT * FROM TerraExecutor_000000369\n               WHERE consumed IS NULL\n               AND   status = 'Succeeded'\n               ORDER BY id ASC\n               LIMIT 1"}
{"severity":"INFO","logging.googleapis.com/sourceLocation":{"file":"wfl/executor.clj","line":386},"timestamp":"2021-11-10T16:57:56.358890Z","message":"jdbc/update! okotsopo@jdbc:postgresql:wfl#612 TerraExecutor_000000369 {:consumed #object[java.time.OffsetDateTime 0x5a569170 \"2021-11-10T16:57:56.358606Z\"], :updated #object[java.time.OffsetDateTime 0x5a569170 \"2021-11-10T16:57:56.358606Z\"]} [\"id = ?\" 1]"}
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":206},"timestamp":"2021-11-10T16:57:56.361499Z","message":{"action":"Sunk workflow outputs","workflow":"668e3257-bf67-41cc-9e77-8be40643c521","entity":["Imputation_Outputs_Table","emerge_test_WFL_identifier_as_input_variable"]}}
{"severity":"INFO","logging.googleapis.com/sourceLocation":{"file":"wfl/sink.clj","line":210},"timestamp":"2021-11-10T16:57:56.367086Z","message":"jdbc/insert okotsopo@jdbc:postgresql:wfl#612 TerraWorkspaceSink_000000369 {:entity \"emerge_test_WFL_identifier_as_input_variable\", :updated #object[java.time.OffsetDateTime 0x12f025cf \"2021-11-10T16:57:56.366850Z\"], :workflow \"668e3257-bf67-41cc-9e77-8be40643c521\"}"}
```
